File: blink-macos/BlinkCore/Sources/BlinkCore/FlowDetection/BreakDecisionEngine.swift
import Foundation

/// Decides whether to show a break overlay or extend the timer.
/// Called when the 20-minute timer fires — evaluates 20 minutes of collected signals.
///
/// This replaces the continuous flow state machine approach.
/// Instead of detecting flow early and extending the timer, we collect
/// signals for the full 20 minutes and make ONE decision at break time.
public final class BreakDecisionEngine {

    /// The decision: interrupt or extend?
    public enum Decision: Equatable, Sendable {
        /// User is doing focused work — extend timer, show gentle nudge
        case extend(minutes: Int, reason: String)
        /// User is casually browsing — show break overlay
        case showBreak
    }

    /// Raw signal window collected over the timer period
    public struct SignalWindow {
        public var keystrokeCount: Int = 0
        public var clickCount: Int = 0
        public var scrollCount: Int = 0
        public var appSwitchCount: Int = 0
        public var windowSeconds: TimeInterval = 0  // how long this window covers
        public var currentAppBundleID: String?

        // Density metrics (computed)
        public var keystrokesPerMinute: Double {
            guard windowSeconds > 0 else { return 0 }
            return Double(keystrokeCount) / (windowSeconds / 60.0)
        }

        public var clicksPerMinute: Double {
            guard windowSeconds > 0 else { return 0 }
            return Double(clickCount) / (windowSeconds / 60.0)
        }

        public var scrollsPerMinute: Double {
            guard windowSeconds > 0 else { return 0 }
            return Double(scrollCount) / (windowSeconds / 60.0)
        }

        public var appSwitchesPerMinute: Double {
            guard windowSeconds > 0 else { return 0 }
            return Double(appSwitchCount) / (windowSeconds / 60.0)
        }

        public init() {}
    }

    // Accumulated signals since last break/reset
    private var window = SignalWindow()
    private var windowStartTime: TimeInterval?
    private var extensionCount: Int = 0  // how many times we've extended (0, 1, 2 max)

    /// Sensitivity (0.4–0.9) — higher = more likely to extend
    public var sensitivity: Double = 0.7

    public init() {}

    // MARK: - Ingest signals

    public func recordKeystroke() {
        window.keystrokeCount += 1
    }

    public func recordClick() {
        window.clickCount += 1
    }

    public func recordScroll() {
        window.scrollCount += 1
    }

    public func recordAppSwitch(bundleID: String) {
        window.appSwitchCount += 1
        window.currentAppBundleID = bundleID
    }

    public func setCurrentApp(bundleID: String) {
        window.currentAppBundleID = bundleID
    }

    /// Call on each tick to update the window duration
    public func tick(now: TimeInterval) {
        if windowStartTime == nil {
            windowStartTime = now
        }
        window.windowSeconds = now - (windowStartTime ?? now)
    }

    // MARK: - Decide

    /// Called when the timer fires. Evaluates all collected signals.
    /// Returns whether to extend the timer or show a break.
    public func decide() -> Decision {
        let kpm = window.keystrokesPerMinute
        let cpm = window.clicksPerMinute
        let spm = window.scrollsPerMinute
        let switches = window.appSwitchesPerMinute
        let isCreativeApp = Self.isCreativeApp(window.currentAppBundleID)

        // How many extensions already used
        if extensionCount >= 2 {
            // Max 2 extensions (20 → 30 → 40 min). No more.
            return .showBreak
        }

        // Score the work intensity (0.0–1.0)
        let score = computeScore(
            kpm: kpm, cpm: cpm, spm: spm,
            switches: switches, isCreativeApp: isCreativeApp
        )

        // Threshold based on sensitivity
        // High sensitivity (0.9) → lower threshold (0.3) → easier to extend
        // Low sensitivity (0.4) → higher threshold (0.7) → harder to extend
        let threshold = 1.1 - sensitivity  // 0.4→0.7, 0.7→0.4, 0.9→0.2

        if score >= threshold {
            extensionCount += 1
            let extendTo = extensionCount == 1 ? 30 : 40
            let reason = describeScore(kpm: kpm, cpm: cpm, isCreativeApp: isCreativeApp)
            return .extend(minutes: extendTo, reason: reason)
        } else {
            return .showBreak
        }
    }

    // MARK: - Score computation

    /// Compute a work intensity score from 0.0 (idle/casual) to 1.0 (deep work)
    private func computeScore(
        kpm: Double, cpm: Double, spm: Double,
        switches: Double, isCreativeApp: Bool
    ) -> Double {
        var score = 0.0

        // Keyboard is the strongest signal (weight: 40%)
        // 0 kpm = 0, 30+ kpm = full score, 80+ kpm = bonus
        let keyboardScore: Double
        switch kpm {
        case 0: keyboardScore = 0
        case ..<10: keyboardScore = 0.2
        case ..<30: keyboardScore = 0.5
        case ..<80: keyboardScore = 0.8
        default: keyboardScore = 1.0
        }
        score += keyboardScore * 0.40

        // Clicks indicate engagement (weight: 20%)
        // Designers click a LOT (30+ cpm). Casual users click 2-5 cpm.
        let clickScore: Double
        switch cpm {
        case 0: clickScore = 0
        case ..<2: clickScore = 0.2
        case ..<5: clickScore = 0.4
        case ..<15: clickScore = 0.7
        default: clickScore = 1.0
        }
        score += clickScore * 0.20

        // Low app switching = focused (weight: 20%)
        // 0-1 switches/min = focused. 3+ = distracted.
        let switchScore: Double
        switch switches {
        case 0: switchScore = 1.0
        case ..<0.5: switchScore = 0.8
        case ..<1: switchScore = 0.5
        case ..<2: switchScore = 0.3
        default: switchScore = 0
        }
        score += switchScore * 0.20

        // Creative app bonus (weight: 10%)
        score += (isCreativeApp ? 1.0 : 0.3) * 0.10

        // Scroll-only penalty (weight: 10%)
        // High scroll + no keyboard = consumption, not creation
        let scrollOnlyScore: Double
        if kpm < 5 && spm > 5 {
            scrollOnlyScore = 0  // scrolling feed, no typing
        } else {
            scrollOnlyScore = 0.5
        }
        score += scrollOnlyScore * 0.10

        return min(score, 1.0)
    }

    private func describeScore(kpm: Double, cpm: Double, isCreativeApp: Bool) -> String {
        if kpm > 30 {
            return "Active typing detected"
        } else if cpm > 10 {
            return "High interaction detected"
        } else if isCreativeApp {
            return "Working in a creative app"
        } else {
            return "Sustained activity detected"
        }
    }

    // MARK: - App classification

    private static let creativeAppPrefixes: Set<String> = [
        "com.microsoft.VSCode",
        "com.apple.dt.Xcode",
        "dev.zed",
        "com.sublimetext",
        "com.jetbrains",
        "com.figma",
        "com.adobe.Photoshop",
        "com.adobe.Illustrator",
        "com.apple.FinalCut",
        "com.apple.Logic",
        "com.googlecode.iterm2",
        "com.apple.Terminal",
        "net.kovidgoyal.kitty",
        "com.github.wez.wezterm",
        "io.alacritty",
        "com.notion",
        "com.obsidian",
        "com.linear",
    ]

    static func isCreativeApp(_ bundleID: String?) -> Bool {
        guard let id = bundleID else { return false }
        return creativeAppPrefixes.contains { id.hasPrefix($0) }
    }

    // MARK: - Reset

    /// Reset after a break is taken or dismissed. Call when timer restarts.
    public func resetWindow() {
        window = SignalWindow()
        windowStartTime = nil
    }

    /// Full reset including extension count. Call on idle/walk-away.
    public func resetAll() {
        resetWindow()
        extensionCount = 0
    }

    /// Current window for debug/display
    public var currentWindow: SignalWindow { window }

    /// How many times we've extended
    public var currentExtensionCount: Int { extensionCount }
}

File: blink-windows/src/Blink.Core/FlowDetection/BreakDecisionEngine.cs
using System.Globalization;

namespace Blink.Core.FlowDetection;

/// <summary>
/// Decides whether to show a break or extend the timer.
/// Called when the 20-minute timer fires. Evaluates accumulated signals.
///
/// Port of Swift BlinkCore.BreakDecisionEngine. Windows matches process
/// NAMES (matched via Contains, case-insensitive) where macOS matches bundle
/// IDs (matched via hasPrefix). The scoring curves, thresholds, dwell
/// tracking and SpotCheck output are kept byte-for-byte equivalent.
/// </summary>
public sealed class BreakDecisionEngine
{
    private int _keystrokeCount;
    private int _clickCount;
    private int _scrollCount;
    private int _appSwitchCount;
    private double _windowSeconds;
    private double? _windowStartTime;
    private int _extensionCount;

    // Per-app dwell time accumulated over this window. Updated by
    // RecordAppSwitch (finalizes previous app) and by FinalizeCurrentApp
    // (called from Decide() to credit the still-frontmost app up to now).
    // Replaces the snapshot-at-decision approach so the creative-app bonus
    // reflects time actually spent, not just whatever app happened to be
    // frontmost when the timer fired.
    private readonly Dictionary<string, double> _dwellByApp = new();
    private string? _currentAppBundleID;
    private double? _currentAppStartTime;

    /// <summary>
    /// Sensitivity (0.4–0.9) — higher = more likely to extend. Required at
    /// construction; BlinkCore has no default.
    /// </summary>
    public double Sensitivity { get; set; }

    public BreakDecisionEngine(double sensitivity)
    {
        Sensitivity = sensitivity;
    }

    // MARK: - Ingest signals

    public void RecordKeystroke() => _keystrokeCount++;
    public void RecordClick() => _clickCount++;
    public void RecordScroll() => _scrollCount++;

    public void RecordAppSwitch(string processName, double? now = null)
    {
        var nowTime = now ?? NowSeconds();
        // Credit the previous app's dwell up to this moment.
        if (_currentAppBundleID != null && _currentAppStartTime.HasValue)
            AddDwell(_currentAppBundleID, Math.Max(0, nowTime - _currentAppStartTime.Value));
        _appSwitchCount++;
        _currentAppBundleID = processName;
        _currentAppStartTime = nowTime;
    }

    public void SetCurrentApp(string processName, double? now = null)
    {
        var nowTime = now ?? NowSeconds();
        // Same dwell-tracking semantics as RecordAppSwitch but without
        // incrementing the switch counter — used at session start.
        if (_currentAppBundleID != null && _currentAppStartTime.HasValue)
            AddDwell(_currentAppBundleID, Math.Max(0, nowTime - _currentAppStartTime.Value));
        _currentAppBundleID = processName;
        _currentAppStartTime = nowTime;
    }

    /// <summary>
    /// Finalize the currently-frontmost app's dwell up to <paramref name="now"/>
    /// and reset the start time so further finalize calls don't double-count.
    /// </summary>
    private void FinalizeCurrentApp(double now)
    {
        if (_currentAppBundleID == null || !_currentAppStartTime.HasValue) return;
        AddDwell(_currentAppBundleID, Math.Max(0, now - _currentAppStartTime.Value));
        _currentAppStartTime = now;
    }

    private void AddDwell(string app, double seconds)
    {
        _dwellByApp[app] = (_dwellByApp.TryGetValue(app, out var v) ? v : 0) + seconds;
    }

    /// <summary>Call on each tick to update the window duration.</summary>
    public void Tick(double now)
    {
        _windowStartTime ??= now;
        _windowSeconds = now - (_windowStartTime ?? now);
    }

    // MARK: - Density metrics

    private double KeystrokesPerMinute => _windowSeconds > 0 ? _keystrokeCount / (_windowSeconds / 60.0) : 0;
    private double ClicksPerMinute => _windowSeconds > 0 ? _clickCount / (_windowSeconds / 60.0) : 0;
    private double ScrollsPerMinute => _windowSeconds > 0 ? _scrollCount / (_windowSeconds / 60.0) : 0;
    private double AppSwitchesPerMinute => _windowSeconds > 0 ? _appSwitchCount / (_windowSeconds / 60.0) : 0;

    /// <summary>
    /// Fraction of total dwell time spent in creative apps. 0 = none, 1 =
    /// entirely creative. Used to scale the creative-app bonus continuously.
    /// </summary>
    private double CreativeFraction(Func<string, bool> isCreative)
    {
        var total = _dwellByApp.Values.Sum();
        if (total <= 0) return 0;
        var creative = _dwellByApp.Where(kv => isCreative(kv.Key)).Sum(kv => kv.Value);
        return creative / total;
    }

    /// <summary>Total seconds spent in creative apps this window.</summary>
    private double CreativeSeconds(Func<string, bool> isCreative)
        => _dwellByApp.Where(kv => isCreative(kv.Key)).Sum(kv => kv.Value);

    // MARK: - Decide

    /// <summary>
    /// Called when the timer fires. Evaluates all collected signals.
    /// <paramref name="maxExtensions"/>: 0 = never extend, 1 = Balanced, 2 = Deep Work.
    /// <paramref name="now"/> is used to finalize the currently-frontmost app's
    /// dwell so the creative-app bonus reflects total time spent.
    /// </summary>
    public BreakDecision Decide(int maxExtensions = 2, double? now = null)
    {
        var nowTime = now ?? NowSeconds();
        FinalizeCurrentApp(nowTime);

        var kpm = KeystrokesPerMinute;
        var cpm = ClicksPerMinute;
        var spm = ScrollsPerMinute;
        var switches = AppSwitchesPerMinute;
        var creativeFraction = CreativeFraction(IsCreativeApp);

        // No extensions allowed (Eye Health) or already used all extensions
        if (maxExtensions == 0 || _extensionCount >= maxExtensions)
            return new BreakDecision.ShowBreak();

        var score = ComputeScore(kpm, cpm, spm, switches, creativeFraction);
        var threshold = 1.1 - Sensitivity;

        if (score >= threshold)
        {
            _extensionCount++;
            var minutes = _extensionCount == 1 ? 30 : 40;
            return new BreakDecision.Extend(minutes, "Focused");
        }

        return new BreakDecision.ShowBreak();
    }

    // MARK: - Score computation
    //
    // Six-bucket keyboard curve (post-2026-05 retune v2). See the macOS
    // BreakDecisionEngine.swift for the full curve-shape rationale + history.
    //   Keyboard rhythm        40%   primary signal
    //   Click rhythm           20%   secondary engagement
    //   App-switch frequency   20%   low switching = focused
    //   Creative-app bonus     10%   IDE / Figma / terminal / etc.
    //   Scroll-vs-work signal  10%   penalizes scroll-dominant patterns

    private static double ComputeScore(double kpm, double cpm, double spm, double switches, double creativeFraction)
    {
        double score = 0;

        // Keyboard (40%) — six-bucket curve.
        double keyboardScore = kpm switch
        {
            0 => 0,
            < 15 => 0.20,   // barely typing
            < 30 => 0.35,   // messaging / slow typing
            < 50 => 0.55,   // active but not deep ← 33 kpm
            < 75 => 0.80,   // steady working
            < 100 => 0.95,  // focused coding
            _ => 1.0,       // deep flow / heads-down
        };
        score += keyboardScore * 0.40;

        // Clicks (20%) — saturate below 1.0 so clicks alone don't dominate.
        double clickScore = cpm switch
        {
            0 => 0,
            < 3 => 0.2,
            < 10 => 0.4,
            < 25 => 0.6,
            _ => 0.8,
        };
        score += clickScore * 0.20;

        // App switching (20%, inverted) — low switching = focused.
        double switchScore = switches switch
        {
            0 => 1.0,
            < 0.5 => 0.8,
            < 1 => 0.5,
            < 2 => 0.3,
            _ => 0,
        };
        score += switchScore * 0.20;

        // Creative app bonus (10%) — continuous, dwell-weighted. Linear
        // interpolation between 0.3 (entirely non-creative) and 1.0 (entirely
        // creative) based on the fraction of dwell time in creative apps.
        score += (0.3 + 0.7 * creativeFraction) * 0.10;

        // Scroll-vs-work signal (10%) — penalize scroll-dominant patterns.
        double scrollWorkScore;
        if (kpm < 5 && spm >= 5)
            scrollWorkScore = 0.0;                          // pure scrolling feed
        else if (spm > kpm * 1.5 && spm > 20)
            scrollWorkScore = 0.2;                          // scroll-dominant
        else if (spm > 30)
            scrollWorkScore = 0.4;                          // high scroll, balanced typing
        else
            scrollWorkScore = 0.8;                          // normal pattern
        score += scrollWorkScore * 0.10;

        return Math.Min(score, 1.0);
    }

    // MARK: - App classification

    /// <summary>
    /// Process names where being frontmost contributes the creative-app bonus.
    /// Matched via Contains (OrdinalIgnoreCase). Windows uses process names,
    /// NOT macOS bundle IDs.
    ///
    /// Intentionally excluded: AI chat clients (Claude Desktop, ChatGPT). Per
    /// product decision 2026-05, agent-driven sessions are reading-heavy and
    /// SHOULD let breaks fire; they don't get the creative bonus.
    /// </summary>
    private static readonly HashSet<string> CreativeApps = new(StringComparer.OrdinalIgnoreCase)
    {
        // VS Code family
        "Code",            // VS Code / VS Code Insiders
        "devenv",          // Visual Studio

        // AI-first editors
        "Cursor",
        "Windsurf",
        "Zed",
        "Trae",

        // Sublime
        "sublime_text",

        // JetBrains family
        "idea64",          // IntelliJ
        "rider64",         // Rider
        "webstorm64",      // WebStorm
        "pycharm64",       // PyCharm
        "clion64",         // CLion
        "goland64",        // GoLand
        "phpstorm64",      // PhpStorm
        "datagrip64",      // DataGrip
        "rubymine64",      // RubyMine
        "fleet",           // JetBrains Fleet
        "studio64",        // Android Studio

        // Design
        "figma",
        "Photoshop",
        "Illustrator",
        "Premiere",        // Adobe Premiere Pro
        "AfterFX",         // Adobe After Effects

        // Terminals — where vim/neovim/emacs/helix users live
        "WindowsTerminal",
        "powershell",
        "pwsh",
        "cmd",
        "alacritty",
        "wezterm-gui",

        // Apple media (shared product list)
        "FinalCut",
        "Logic",

        // Knowledge work
        "Notion",
        "Obsidian",
        "Linear",
    };

    private static bool IsCreativeApp(string? name) =>
        name != null && CreativeApps.Any(c => name.Contains(c, StringComparison.OrdinalIgnoreCase));

    // MARK: - Reset

    /// <summary>
    /// Reset after a break is taken or dismissed. Preserves the currently-
    /// frontmost app (re-seeds dwell tracking from <paramref name="now"/>) so
    /// the caller need not re-send an app-switch the user didn't trigger.
    /// </summary>
    public void ResetWindow(double? now = null)
    {
        var nowTime = now ?? NowSeconds();
        var preservedApp = _currentAppBundleID;
        _keystrokeCount = 0;
        _clickCount = 0;
        _scrollCount = 0;
        _appSwitchCount = 0;
        _windowSeconds = 0;
        _windowStartTime = null;
        _dwellByApp.Clear();
        _currentAppBundleID = null;
        _currentAppStartTime = null;
        if (preservedApp != null)
        {
            _currentAppBundleID = preservedApp;
            _currentAppStartTime = nowTime;
        }
    }

    /// <summary>Full reset including extension count. Call on idle/walk-away.</summary>
    public void ResetAll(double? now = null)
    {
        ResetWindow(now);
        _extensionCount = 0;
    }

    public int CurrentExtensionCount => _extensionCount;

    // MARK: - Spot check

    /// <summary>
    /// Compute the current score and return a debug summary without affecting
    /// state (no extension count changes). Finalizes the current app's dwell
    /// up to <paramref name="now"/> so creativeFraction is current.
    /// </summary>
    public SpotCheckResult SpotCheck(double? now = null)
    {
        var nowTime = now ?? NowSeconds();
        FinalizeCurrentApp(nowTime);
        var kpm = KeystrokesPerMinute;
        var cpm = ClicksPerMinute;
        var spm = ScrollsPerMinute;
        var switches = AppSwitchesPerMinute;
        var creativeFraction = CreativeFraction(IsCreativeApp);
        var creativeSeconds = CreativeSeconds(IsCreativeApp);
        var score = ComputeScore(kpm, cpm, spm, switches, creativeFraction);
        var threshold = 1.1 - Sensitivity;
        return new SpotCheckResult(
            score, threshold, score >= threshold,
            kpm, cpm, spm, switches,
            creativeFraction, creativeSeconds,
            _currentAppBundleID, _windowSeconds, _extensionCount);
    }

    private static double NowSeconds() =>
        ((DateTimeOffset)DateTime.UtcNow).ToUnixTimeMilliseconds() / 1000.0;

    /// <summary>
    /// Snapshot of the flow score + signals, used by the app's debug toast.
    /// Port of Swift BreakDecisionEngine.SpotCheck.
    /// </summary>
    public readonly record struct SpotCheckResult(
        double Score,
        double Threshold,
        bool WouldExtend,
        double Kpm,
        double Cpm,
        double Spm,
        double AppSwitchesPerMinute,
        double CreativeFraction,
        double CreativeSeconds,
        string? CurrentApp,
        double WindowSeconds,
        int ExtensionCount)
    {
        private static string F(double v, string fmt) => v.ToString(fmt, CultureInfo.InvariantCulture);

        /// <summary>One-liner for the menu bar.</summary>
        public string Summary =>
            $"Score: {F(Score, "0.00")} (threshold: {F(Threshold, "0.00")}) → {(WouldExtend ? "EXTEND" : "BREAK")}";

        /// <summary>Full detail for preferences/logs.</summary>
        public override string ToString()
        {
            var verdict = WouldExtend ? "EXTEND" : "BREAK";
            var mins = (int)WindowSeconds / 60;
            var secs = (int)WindowSeconds % 60;
            var creativeMins = (int)CreativeSeconds / 60;
            var creativeSecsRem = (int)CreativeSeconds % 60;
            var creativePct = (int)Math.Round(CreativeFraction * 100);
            return
                $"Score: {F(Score, "0.00")} (threshold: {F(Threshold, "0.00")}) → {verdict}\n" +
                $"Window: {mins}m {secs}s | Extensions used: {ExtensionCount}\n" +
                $"Keyboard: {F(Kpm, "0.0")} kpm\n" +
                $"Clicks: {F(Cpm, "0.0")} cpm\n" +
                $"Scrolls: {F(Spm, "0.0")} spm\n" +
                $"App switches: {F(AppSwitchesPerMinute, "0.0")}/min\n" +
                $"Creative time: {creativeMins}m {creativeSecsRem}s ({creativePct}%)\n" +
                $"App (current): {CurrentApp ?? "none"}";
        }
    }
}

// Decision types
public abstract record BreakDecision
{
    public sealed record ShowBreak() : BreakDecision;
    public sealed record Extend(int Minutes, string Reason) : BreakDecision;
}
